bdk_core/src/sparse_chain.rs

```diff
@@ -226,447 +139,229 @@ impl SparseChain {
+                    .collect::<Vec<_>>();
+                debug_assert!(txids.len() <= 1, "conflicting txs in sparse chain");
+                txids.get(0).cloned()
+            })
+            .flatten()
+            .cloned();
 
         Some(FullTxOut {
-            value: txout.value,
-            spent_by,
             outpoint,
-            script_pubkey: txout.script_pubkey.clone(),
-            confirmed_at: tx_in_block.confirmation_time,
+            txout,
+            height,
+            spent_by,
         })
     }
 
-    /// Iterates over all transactions related to the descriptor ordered by decending confirmation
-    /// with those transactions that are unconfirmed first.
-    ///
-    /// "related" means that the transactoin has an output with a script pubkey produced by the
-    /// descriptor or it spends from such an output.
-    // TODO:  maybe change this so we can iterate over a height range using checkpoints
-    pub fn iter_tx_by_confirmation_time(
-        &self,
-    ) -> impl DoubleEndedIterator<Item = (Txid, &TxAtBlock)> + '_ {
-        // Since HashSet is not necessarily a DoubleEndedIterator we collect into a vector first.
-        let mempool_tx = self
-            .mempool
-            .iter()
-            .map(|txid| (*txid, self.txs.get(txid).unwrap()))
-            .collect::<Vec<_>>();
-        let confirmed_tx = self.checkpoints.iter().rev().flat_map(|(_, data)| {
-            data.ordered_txids
-                .iter()
-                .map(|(_, txid)| (*txid, self.txs.get(txid).unwrap()))
-        });
-
-        mempool_tx.into_iter().chain(confirmed_tx)
+    pub fn set_checkpoint_limit(&mut self, limit: Option<usize>) {
+        self.checkpoint_limit = limit;
     }
 
-    pub fn unconfirmed(&self) -> &HashSet<Txid> {
-        &self.mempool
-    }
-
-    /// Get a stored transaction given its `Txid`.
-    pub fn get_tx(&self, txid: Txid) -> Option<&TxAtBlock> {
-        self.txs.get(&txid)
-    }
-
-    /// internal debug function to double check correctness of the accumulated digest at the tip
-    #[must_use]
-    fn is_latest_checkpoint_hash_correct(&self) -> bool {
-        if let Some(tip) = self.latest_checkpoint() {
-            let mut txs = self
-                .iter_tx()
-                .filter(|(_, tx)| tx.confirmation_time.is_some())
-                .collect::<Vec<_>>();
-            txs.sort_by_key(|(_, tx)| (tx.confirmation_time.unwrap().height, tx.tx.txid()));
-            let mut hasher = sha256::HashEngine::default();
-            for (txid, _) in txs {
-                hasher.input(&txid);
-            }
-            let txid_hash = sha256::Hash::from_engine(hasher);
-            self.txid_digest_at(tip) == txid_hash
-        } else {
-            true
-        }
+    pub fn prune_checkpoints(&mut self) -> Option<BTreeMap<u32, BlockHash>> {
+        let height = *self.checkpoints.keys().rev().nth(self.checkpoint_limit?)?;
+        return Some(self.checkpoints.split_off(&height));
```

Comment:
This logic seems wrong. If my checkpoint limit is `Some(0)` then it should drain all the checkpoints (0 checkpoints are allowed). If it's `Some(1)` only the most recent one should be kept. 

`split_off` is splitting the more recent ones though :S

Reply:
Should be fixed, need to test.